File: 896-monotonic-array/896-monotonic-array.py
# O(n) time | O(1) space
class Solution:
    def isMonotonic(self, nums: List[int]) -> bool:
        # handle edge case
        if len(nums) <= 2:
            return True
        # determine whether array is decreasing or increasing
        dir = nums[1] - nums[0]
        startIdx = 2
        for idx in range(2, len(nums)):
            if dir == 0:
                dir = nums[idx] - nums[idx - 1]
                startIdx = idx
                continue
            break
        # check for the entire array
        for idx in range(startIdx, len(nums)):
            if dir > 0 and not nums[idx] >= nums[idx - 1]: return False
            elif dir < 0 and not nums[idx] <= nums[idx - 1]: return False
        return True

File: 896-monotonic-array/896-monotonic-array.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
// O(n) time | O(1) space
var isMonotonic = function(nums) {
    // handle edge case
    if (nums.length <= 2) return true;
    // determine whether the array is decreasing or increasing
    let dir = nums[1] - nums[0];
    let startIdx = 2;
    for (let i = 2; i < nums.length; i++){
        if (dir !== 0) break
        dir = nums[i] - nums[i - 1]
        startIdx = i
    }
    // check for the entire array
    for (let i = startIdx; i < nums.length; i++){
        if (dir > 0 && !(nums[i] >= nums[i-1])) return false
        if (dir < 0 && !(nums[i] <= nums[i-1])) return false
    }
    return true
};